README

PHASEX-0.14.98-dev

-------------------------------------------------------------------------------
[P]hase [H]armonic [A]dvanced [S]ynthesis [EX]periment
-------------------------------------------------------------------------------

Welcome to the latest and greatest of PHASEX!

PHASEX is an experimental MIDI softsynth for Linux/ALSA/JACK with a
synth engine built around flexible phase modulation and flexible
oscillator/LFO sources.  Modulations include AM, FM, offset PM, and
wave select.  PHASEX comes equipped with multiple filter types and
modes, a stereo crossover delay and chorus with phaser, ADSR envelopes
for both amplifier and filter, realtime audio input processing
capabilities, and more.  Inspirations come from a variety of analog
and early digital MIDI synthesizers from the '80s and '90s.

-------------------------------------------------------------------------------


GOALS:
-------------------------------------------------------------------------------

* C Hacker Friendly.  PHASEX is written entirely in C.  No C++.  No
  assembly.  From the very beginning, the program framework (now
  a much larger application framework) has existed to support a synth
  engine written in good ol' C.  Now the application framework is well
  on its way to serving as a basic platform for developing new
  instruments written in... you guessed it... C.

* Reliability.  Whether playing music on stage, working in the studio,
  or writing music on a laptop in a coffee shop, the instruments and
  tools need to function properly and consistently in order to achieve
  good results.  Musicians need to be able to focus on the music, not
  shortcomings of the toolset.

* Compatibility, starting with full JACK and ALSA API support (and
  eventually support for additional audio / MIDI libraries), allowing
  PHASEX to be used in any pro audio workflow on a modern Linux
  distribution.  The standalone application design was partially
  chosen for this reason -- not all sequencers and DAWs support all
  instrument plugin formats.  For this reason, PHASEX will most likely
  see support for features such as ALSA channel mapping, JACK
  Freewheel, and even Non Session Manager (NSM) before seeing support
  for DSSI or LV2?  Why?  PHASEX already runs side by side with
  Ardour, MusE, Rosegarden, Qtractor, Seq24, or any other sequencer
  supporting JACK MIDI or the ALSA sequencer interface, no matter what
  plugin APIs these sequencers do or do not happen to support.  In the
  future, adding support for plugin formats may be deemed necessary
  for making PHASEX more useful in certain situations.  For now, the
  development effort will focus on fine-tuning and extending the usage
  of the APIs already supported.

* Stage friendly.  Low MIDI latency and low timing jitter.  Run with
  or without the GUI.  Ability to specify alternate config files
  (just be sure you have a backup of your ~/.phasex/on-stage.cfg...).
  With decent hardware running PHASEX and a recent -rt kernel with
  proper permissions, priority tuning, hardware tuning (BIOS settings,
  PCI latencies, etc.), and audio buffer tuning, MIDI-audio latencies
  of under 10ms should be well within reach for audio hardware capable
  of running with 128 sample buffer periods or better.

* Studio friendly.  Focus on workflow.  The recent development work
  has been aimed at making PHASEX easier to use and efficient to
  manage (quick configuration, built in session management, MIDI
  connections in the menus, etc.).  Again, the focus should be on the
  music, not on remembering how to make the instrument jump through a
  bunch of hoops just to do it's job.  PHASEX development will
  continue in this direction with new session management features
  based on community feedback.

* Great sound.  This one is of course, quite subjective, but it still
  comes down to the same juggling act on the programming end:  Choosing
  the "right" algorithms, making resource usage vs. sound quality
  decisions, calculation speed vs. accuracy decisions, and of course,
  decisions about how much programming time is spent on what features
  (hmmm... is it worth the time to do the math and write and debug the
  code for nifty-feature-x, or is time better spent fixing
  broken-feature-y?).  At this point, a lot of these decisions have
  been made and the synth engine as a whole is very stable, but these
  things are not necessarily set in stone.  There is always room for
  improvement by researching new ways of doing the math and learning
  new DSP tricks for solving some of the flaws inherent in the current
  methods.  So, how does it sound?  Decide for yourself.


FEATURES:
-------------------------------------------------------------------------------

* GTK Interface:

    PHASEX has a GUI with switchable color themes, knobs and buttons
    for all 150+ patch parameters, preferences dialog for easy
    configuration, patch bank and session bank, MIDI connections from
    the main menu, and more.  True fullscreen mode, notebook,
    one-page, and widescreen layouts, and select-able fonts and knob
    sizes allow PHASEX to fit into any amount of screen real-estate
    (now able to squeeze into sizes smaller than 800x600, or more
    comfortably sit in a 1920x1080 desktop).

* Polyphonic:

    PHASEX is polyphonic.  In poly mode, voices are allocated
    dynamically up to the runtime configurable voice limit.  
    Oscillators and filters function independently per-voice.  LFOs,
    however, function on a per-part basis.  In addition to Poly mode,
    the three legacy Mono modes are still supported: Mono Smooth mode
    puts new notes back into sustain if playing legato.  Mono Multikey
    mode maps the keys played, in order, round-robin style to all
    oscillators based on MIDI note frequencies in use.  In Mono
    Retrigger mode, envelopes and oscillator init phases are
    retriggered every time a note is played.

* Multitimbral:

    Many parts of PHASEX were re-architected during the v0.14.x
    development phase.  This release is now fully multitimbral, with
    each part running in its own thread.  The number of parts (and
    thus engine threads) is determined at compile time, and has been
    tested rigorously with 8 voices on a quad-core q6600.  Part
    switching in the GUI is clean and efficient.  Session handling
    support takes the fuss out of keeping track of session patches.

* ALSA Audio:

    PHASEX supports native ALSA for flawless xrun free audio playback.
    Sample rate and buffer size are easily configurable to achieve the
    lowest possible latencies for your system.

* JACK Audio:

    PHASEX has supported JACK audio from the beginning of the public
    MIDI enabled releases.  Any sample rate or buffer size used by
    jackd can be utilized.  Both single- and multi- stereo-pair
    routings are supported.  Basic Transport is supported, but as of
    this time, the tempo and LFO sync are largely untested and may
    still need some fine-tuning.

* ALSA MIDI:

    The ALSA sequencer interface has been with PHASEX since its first
    public release.  As of v0.14.x, the ALSA raw MIDI interface is
    also supported, offering lower timing jitter when controlling
    PHASEX via hardware MIDI.

* JACK MIDI:

    JACK MIDI provides 100% sample-accurate timing by design.  PHASEX
    has embraced this design, and the new underlying MIDI system is
    the result.

* Generic MIDI:

    Generic MIDI (/dev/midi) support is provided for communicating
    with MIDI devices or drivers that use a standard POSIX interface
    supporting open/close/poll/read/fcntl on a named pipe, socket, or
    device file.  The generic MIDI support has only been tested with
    ALSA's /dev/midi* and /dev/snd/midiC?D? device files (which are
    accessed through the POSIX interface, not the alsa-lib API), but
    should perform the same with any MIDI device interface as long as
    additional system and library calls are not needed.  Currently,
    compatibility with serial to MIDI converters is unknown, so please
    send feedback (positive or negative) if you attempt to use a
    serial to MIDI converter.

* MIDI Control:

    All parameters are quickly mappable and ready for full realtime
    control.  Parameter names are right clickable for quick and easy
    MIDI controller assignments.  Each MIDI controller can be mapped
    to up to 15 parameters.  Velocity sensitivity and key aftertouch
    can modulate the amp and the filter, or used as raw oscillators of
    LFOs.  MIDI maps may be saved and later restored to facilitate
    switching between different working environments.

* Realtime MIDI Clock:

    Addresses the issue of how to accurately synchronize MIDI events
    with individual frame positions within a buffer period instead of
    quantizing the timing to whatever the audio buffer period happens
    to be.  Typical audio hardware supports stable realtime operation
    at anywhere from 2-20 milliseconds latency.  With no additional
    synchronization, this audio latency becomes a self-imposed
    quantization factor on event timing, adding MIDI timing jitter
    roughly equal to the duration of one audio buffer processing
    period, hence the need for a MIDI clock:

    A timerless, tickless audio / MIDI timing synchronization system.
    The clock latches onto the audio buffer processing periods and
    generates a high-precision MIDI clock reference, with much less
    jitter than timing based on when the audio processing cycle
    actually wakes up (which can really be anywhere within the
    period).  This clock is agnostic to the specific audio and MIDI
    drivers being used and requires no timer interrupts.  In essence,
    the timing accuracy achieved is only dependent on the MIDI
    hardware/software and realtime scheduling latencies of the system.
    With decent professional MIDI and audio hardware, sub-millisecond
    timing accuracy is available with total MIDI event latency under
    10ms, accurate and responsive enough for most professional sound
    sound engineers and musicians.

* Multithreaded:

    PHASEX uses separate threads with realtime scheduling for audio
    output, MIDI input, and the synth engine.  The GTK UI runs in its
    own thread at normal scheduling priority.

* 4 Oscillators:

    Each oscillator features wave shape selection (including sampled
    and generated waveforms), bipolar or unipolar output, frequency
    source selection (midi note, tempo based, tempo based w/ phase
    trigger, audio input, or velocity/aftertouch), mix modulation
    supporting standard or AM mixing, per-osc transpose and
    pitchbender amounts, and select-able modulation sources for AM, FM,
    phase offset, and waveshape modulation.  For all but waveshape
    modulation, both oscillators and LFOs are available as modulation
    sources.

* 4 LFOs:

    The dedicated per-osc LFOs have been replaced with generic
    per-part LFOs.  The LFOs are identical to the oscillators with the
    following exceptions:  They are per-part instead of per-voice,
    they do not accept other modulation sources (yet), they are not
    transposed by the master transpose parameter, and when sourced to
    velocity, aftertouch data is ignored.  New routing capabilities
    have been added with the Voice AM, LFO-1 Rate, Cutoff, and LFO-3
    Rate per-LFO dedicated function parameters.

* ADSR Envelopes:

    PHASEX comes equipped with standard ADSR envelopes for the
    amplifier and filter.  Duration times follow a parabolic curve.
    The amplitude of the amplifier envelope follows a logarithmic
    curve.  The filter envelope is linear with respect to the musical
    scale.  Both envelopes can also be used as oscillator and LFO
    sources.

* Filter:

    The first filter style is a Chamberlin filter (2-pole 12db/octave)
    hacked up so it doesn't destabilize at high cutoff or resonance
    values.  Low pass, high pass, band pass, and band stop modes are
    available, as well as four combo modes.  The filter has it's own
    ADSR envelope (identical to amplifier envelope with the added
    option of switching to negative polarity).  One of the LFOs may be
    mapped to modulate filter cutoff.  Cutoff controller values map
    directly to MIDI note numbers.  Optional keyfollow mode allows
    cutoff to follow the key played for the MIDI note, or the lowest,
    highest, or most recent key in play.  The filter is per-voice in
    Poly mode.

    As of v0.14.x, a 4-pole 24db/octave Stilson/Smith style Moog
    filter has been added.  In PHASEX style, cutoff frequency is
    adjusted in halfstep increments, and both a distorted and a
    clean version are available.

    The 3-pole filter is experimental, and might not be included in
    future versions.  Future plans include fixing the frequency
    calculation and turning this into a proper 303 style filter.

* Chorus:

    The PHASEX stereo crossover chorus (inspired by the chorus on the
    old Akai S-series samplers) adds depth and that extra bit of
    stereo separation for an even wider stereo field.

* Delay:

    The delay sports the normal mix and feedback controls, along with
    a channel crossover option.  A dedicated LFO allows modulation of
    playback position, great for flange-like effects (or really
    messing with people's heads at longer delay times).  Delay times
    are select-able based on tempo, in both 1/64th and 1/48th
    increments.

* Input Processing:

    Up to two JACK input channels (as mono, dual mono, or stereo) may
    be used as audio sources for both oscillators and LFOs.  An input
    envelope follower and input boost allow for sources like guitar or
    vocals to be processed with ease.

* Stereo Width:

    The width of the stereo image is adjustable via a MIDI controller
    from pure mono to pure stereo.  Can be used during sound design to
    help create patches that sound very different when a center
    playback channel is present.  Very useful for getting patches to
    sit in the mix properly.

* Velocity Sensitivity:

    Velocity of Note-On events can modify the filter or amplifier, and
    can also be routed through the LFOs.

* Session Handling:

    For multitimbral builds, PHASEX comes equipped with a session bank
    which operates much like the existing patch bank.  In addition to
    the new native session handling, LASH and JACK Session supported
    as well.

-------------------------------------------------------------------------------


MAJOR CHANGES SINCE v0.12.x:
-------------------------------------------------------------------------------

In addition to new major features listed above, all parts of the
PHASEX source code have been overhauled, with some important changes:

* New synth patch file format:  This decision was not made lightly,
  and there are many reasons for this:  The behavior of some
  parameters has changed (anything dealing with velocity curves,
  envelope curves, offset of negative envelopes), new parameters have
  been added, new parameter values have been added, and the old patch
  parser code grew into an almost unmanageable mess.  Starting with
  v0.14.x, parameter values in patch files reflect the values shown in
  the GUI instead of _mostly_ reflectling MIDI values, which should be
  better for human readability.  Improperly named parameters have been
  renamed to avoid confusion as well.  A patch conversion script is
  provided to convert patches from v0.12.x to v0.14.x, and should work
  well in most cases.  With the new synth engine improvements, a lot
  of old patches can be modified to sound better with very little
  effort, so please don't fret when old patches don't sound _exactly_
  the same.  If there is one apology that needs to be made with the
  new PHASEX release, this is it.  If converting all your old patches
  into the new format really throws a monkeywrench into the works, I
  do apologize.

* GUI and engine are (almost) fully separated.  Engine parameters and
  GUI parameters now have separate callback paths.  The GUI queues
  parameter change events for the engine instead of modifying engine
  parameters directly.  GUI-initiated patch changes are still
  performed by the GUI (hence 'almost separated').  Current plans
  include separating this out as well.  Cleaning up this aspect of the
  GUI has helped tremendously, allowing all the weird hacks and
  workarounds for parameter state handling to be replaced with a
  design that works efficiently and dependably.  Since the engine and
  MIDI code no longer have to be aware of the GUI, phasex can now be
  started without the GUI.

* Lock-free buffer synchronization:  The pthreads mutexes in the
  buffer processing code been removed in favor of lock-free buffering
  and for the engine, clock_nanosleep() based adaptive sleep timing.
  Moving away from the old broken design means fewer context switches,
  fewer (if any) priority inversion problems, a better guarantee that
  engine threads can fill buffers on time, and most important of all:
  no more "bad PHASEX sound" from engine and audio losing sync.

* New MIDI queue:  Single-reader.  Multiple-writer.  Multiple events
  per frame.  Monotonic.  Reliable.  Used by GUI and all MIDI drivers
  to queue events for the engine.  Queue data format is designed to
  hold 2 and 3 byte MIDI messages.  Event types not defined in the
  MIDI spec are used for internal non-MIDI events (currently 0x01
  through 0x04).

* MIDI Reference Clock:  All of the MIDI event queuing and realtime
  buffer synchronization between audio, engine, and MIDI threads is
  built around the new MIDI reference clock.  All synchronization
  events (engine waking up, audio waking up, timestamping MIDI events,
  and updating audio,engine, and MIDI buffer indices) happen at a time
  relative to the begininng of a MIDI period (equal in time to an
  average audio buffer period, but starting some configurable amount
  of time earlier).

* There's plenty more if you're curious.  Have a look at the
  ChangeLog, or just go straight to the source.

-------------------------------------------------------------------------------


REQUIREMENTS:
-------------------------------------------------------------------------------

Required:
  * ALSA >= 1.0.18
  * JACK (JACK or JACK2/jackdmp should work).
  * libsamplerate
  * GTK-2 (version 2.8 or later recommended).
  * Realtime Linux Kernel (latest 3.0.x-rt or newer recommended).
  * 1.6 GHz or faster CPU (multiple core, 64-bit preferred).

Optional:
  * LASH >= 0.5.4.
  * libuuid (required by LASH).

-------------------------------------------------------------------------------


COMPILING AND INSTALLING PHASEX:
-------------------------------------------------------------------------------

For a standard, multitimbral configuration optimized for your CPU,
run the following:

    cd phasex
    aclocal && autoconf && automake && autoheader
    ./configure --enable-arch=native --enable-parts=2
    make
    make install

Other useful configure flags are --enable-debug=, --enable-32bit,
--enable-cpu-power=, and --without-lash.

See INSTALL for full compilation and installation instructions.

-------------------------------------------------------------------------------


RUNNING PHASEX:
-------------------------------------------------------------------------------

Usage:  phasex [options] [<patch>[,<patch>[,<patch>[,...]]]]

  -c, --config-file=     Alternate config (takes precedence over default
                             config file and command line options.
  -A, --audio-driver=    Audio driver:  alsa or jack.
  -a, --audio-device=    ALSA device name (hw:x,y format).
  -r, --sample-rate=     Audio sample rate (ALSA only).
  -p, --latency-periods= Number of buffer latency periods (ALSA only).
  -s, --period-size=     Buffer period size (power of 2, ALSA only).
  -M, --midi-driver=     MIDI driver:  jack, alsa-seq, alsa-raw, generic.
  -m, --midi-port=       MIDI input port or device name (driver specific).
  -f, --fullscreen       Start GUI in fullscreen mode.
  -x, --maximize         Start GUI with main window maximized.
  -b, --bpm=             Override BPM in patch bank and lock BPM parameter.
  -t, --tuning=          Base tuning frequency for A4 (default 440 Hz).
  -i, --input=           Comma separated pair of audio input matches.
  -o, --output=          Comma separated pair of audio output matches.
  -O, --oversample       Use double the sample rate for internal math.
  -U, --undersample      Use half the sample rate for internal math.
  -G, --no-gui           Run PHASEX without starting the GUI.
  -D, --session-dir=     Set directory for loading initial session.
  -u, --uuid=            Set UUID for JACK Session handling.
  -d, --debug=           Debug class (Can be repeated. See debug.c).
  -l, --list             Scan and list audio and MIDI devices.
  -v, --version          Display version and exit.
  -h, --help             Display this help message and probe ALSA hardware.

LASH Options:
  -P, --lash-project=    LASH project name.
  -S, --lash-server=     LASH server address.
  -I, --lash-id=         LASH client ID.
  -L, --disable-lash     Disable LASH completely for the current session.

Please note that unless an alternate config file is used, all command
line options take precedence over settings saved in the config file.

If using phasex with JACK, make sure jackd is running (and configured
for your system) before running phasex.  If jackd is not running,
PHASEX will fall back to using the ALSA drivers.  If you use software
like qjackctl to handle your audio and MIDI patchbay, then the command
line port connection options aren't necessary.  The -i and -o options
accept comma separated substring pairs to match on JACK audio port
names.  The -m option takes a comma separated list of MIDI ports (X:Y
for ALSA Seq, hw:X,Y for ALSA Raw, or device file for generic raw
MIDI).  For example:

    phasex -A jack -i _1,_2 -o _1,_2 -M alsa-seq -m 16:0,20:0

will match JACK ports alsa_pcm:capture_1 and alsa_pcm:capture_2 for
input, alsa_pcm:playback_1 and alsa_pcm:playback_2 for output, and the
first two ALSA sequencer hardware MIDI input ports with relatively
recent versions of JACK and ALSA.  (Jack2/jackdmp uses 'system'
instead of 'alsa_pcm'.  ALSA port designations vary from system to
system.)

To view a full technical description of any parameter, simply
middle-click on the parameter label.  Similarly, a right-click on any
parameter label allows on-the-fly mapping of any MIDI controller to
any parameter (up to 15 parameters may be mapped to a single MIDI
controller).  All remaining features are directly visible and
accessible through the GUI and its (very short) menus.

Patch bank storage works like it does on just about any modern
hardware synth, with one exception: Patches themselves are stored by
name in the filesystem, which are then mapped by program number in the
patch bank.  The individual program slots in the patch bank simply
reference the stored patches by name.  In Autosave mode, a patch is
automatically saved before a new patch is loaded during a program
change without any user intervention.  Autosave mode is intended for
use during sound design sessions (and should be avoided the rest of
the time to keep from overwriting patches).  In Warn mode, warnings
will be issued will be issued when unsaved modifications to the patch
will be lost.  In Protect mode, program changes always take effect
immediately, so any patch modifications are quietly discarded and
patches never get (re)saved unless they are saved explicitly.

-------------------------------------------------------------------------------

In the event of catastrophic failure, and PHASEX fails to start, edit
~/.phasex/phasex.cfg and look for the last settings change, or simply
remove phasex.cfg (but not any directories) and PHASEX will start with
the default config and auto-detect your setup all over again.

-------------------------------------------------------------------------------


LICENSING AND COPYRIGHT:
-------------------------------------------------------------------------------

PHASEX is distributed under the terms of the GNU Public License,
version 3.  See LICENSE for details.

    Copyright (C) 1999-2013 William Weston <william.h.weston@gmail.com>.

With portions of and contributions to the PHASEX source:

    Copyright (C) 2010 Anton Kormakov.
    Copyright (C) 2007 Peter Shorthose.
    Copyright (C) 2004 Sean Bolton.
    Copyright (C) 1999 Tony Garnock-Jones.

See AUTHORS for details.


ChangeLog
[P]hase [H]armonic [A]dvanced [S]ynthesis [EX]permient

Changelog:
2015-03-04 William <william.h.weston@gmail.com> - 0.14.98-dev
	* Updated email address.
	* Included Fons Adriaensen's DC filter patch.
	* Fixed memory clobbering bug in sample resampling code.
	* Applied Ralf's DC offset fix for sample normalization.
	* Added analog sine oscillator w/ 2 bandlimiting levels.
	* Fixed improper install of autoconf scripts.
	* Removed -fsingle-precision-constant from OPT_CFLAGS.
	* Created macros for wavetable lookup funtions.
	* Added conditional compilaton for splash screen.
	* Fixed portamento overshoot bug.
	* Increased internal tuning resolution to 1440.
	* Fixed gtkknob compilation warnings.
	* Enabled JACK MIDI with JACK audio, regardless of midi_driver.
	* Added MIDI_EVENT_RESYNC internal MIDI event type.
	* Added new rt-safe thread resync via internal MIDI transport.
	* Reworked startup and sync for threads, buffer indexes, and clock.
	* Fixed startup segfault in set_midi_channel_for_part();
	* Fixed improper GError (non)initilization in multiple places.


2013-01-11 William Weston <william.h.weston@gmail.com> - 0.14.97
	* Cleaned up pointer sign warning in queue_midi_event().
	* Included (/usr)/bin/true configure.ac fix from Jörn Nettingsmeier.
	* Included falktx's build patch for src/Makefile.am.
	* Added synth-part to MIDI channel assignments to midimaps.
	* Removed unused type_id from struct jack_port_info.
	* Added proper detection and conditions for new JACK features.
	* Reworked BPM handling and removed duplicated code.
	* Fixed typo in pixmaps/Makefile.am.  (Adam Sampson)
	* Implemented note-off instant release from attack and decay.
	* Implemented round-robin voice assignment for mono-multikey.
	* Fixed bug with assigning MIDI CC in GUI.
	* Changed MIDI CC mapping event (now ctrl-middle-click).
	* Changed depracated AM_CONFIG_HEADER to AC_CONFIG_HEADERS.
	* Fixed FTBFS on systems w/o CLOCK_MONOTONIC_RAW.  (Adrian Knoth)
	* Fixed crash and clock sync issues with changing jack buffer size.
	* Consolidated gtkknob frame offset calculations.
	* Added checks to prevent errors on JACK client close.
	* Fixed usage of JACK stereo/multi menuitems for monotimbral builds.
	* Fixed potential LASH client name bug when using ALSA MIDI.
	* Added JACK latency recalculation when changing buffer_latency.
	* Added git version tag tracking to build system.
	* Changed default locale to "C" instead using env defaults.
	* Fixed settings not saved on quit.
	* Fixed new dirs/files not created bug in phasex-convert-patch.
	* Fixed GUI param update race condition on updated flag.
	* Fixed sporadic GUI param sensitivity update issue.
	* Updated system synth-patches.
	* Fixed FTBS with hardening flags enabled.  (Alessio Treglia)


2012-12-28 William Weston <william.h.weston@gmail.com> - 0.14.96
	* Added initial JACK Session support.
	* Added sess dump dir skipping for Load Session dialog.
	* Modified MIDI event queue to allow multiple writers.
	* Removed get_/set_midi_event_state().
	* Replaced all_notes_off[] flags with queued events.
	* Implemented event queuing for GUI param changes.
	* Renamed all variables simply named 'index'.
	* Renamed wave_table to more appropriate osc_table.
	* Fixed JACK MIDI port connection tracking.
	* Fixed JACK latency callback calculations.
	* Added managed mode w/ arg to load_/save_session().
	* Cleaned up LASH session load/restore w/ new session mods.
	* Updated configure.ac with new x86 CPU optimizations.
	* Changed autosave sess name from _current_ to _autosave_.
	* Enabled notes off on transport stop when JACK Transport is off.
	* Removed redundant poll() call from rawmidi thread.
	* Fixed MIDI controllers above max bug.
	* Fixed oscillator scaling in osc_table.
	* Completed work for pure 64-bit math (cpu power level 4).
	* More cleanups and depracated code removal.


2012-12-20 William Weston <william.h.weston@gmail.com> - 0.14.95
	* Fixed initial patch / default patch issues.
	* Updated GPL msg on exit to reference GPLv3 instead of v2.
	* Added 16 new patches.
	* Applied Anton Kormakov's steal-oldest-note fix.


2012-12-20 William Weston <william.h.weston@gmail.com> - 0.14.94
	* Major code overhaul since 0.12.
	*   Renamed variables and functions to improve readability.
	*   Rewrote comments where old comments no longer made sense.
	*   Rewrote at least half of the previous codebase.
	*   Overhauled all major systems, as noted below.
	*   Removed thousands of lines of depracated code.
	*   Reworked startup / shutdown code (very stable).
	* Added debug thread and DEBUG/ERROR/WARN macros.
	*   Added debug classes for selective debug output.
	* Build-system overhaul.
	*   Added new optimization flags for better performance.
	*   Fixed library detection (should build on more systems).
	*   Fixed inconsistent installs and uninstalls.
	*   Fixed handling of icons and desktop file.
	*   Updated phasex.spec.in for Fedora 14-18 builds.
	*   Added rpm target to toplevel Makefile.am.
	* Integration of new features from Anton Kormakov (work in progress).
	*   Added LASH session support (stable).
	*   Added hold pedal support (untested).
	*   Added JACK Transport support (currently broken).
	*   Added DC rejection filter in synth engine (for debugging).
	*   Fixed Chorus Time calculations and typo in chorus math.
	* Multitimbral architecture (thread per part).
	*   Data structure overhaul, affecting almost everything.
	*   Added Stereo- and multi- outputs for JACK.
	*   Added GUI part switching with hotkeys (F1-F<part_num>).
	*   Added Session management in with session bank in GUI.
	*   Removed pthreads mutexes in favor of atomic index values.
	* Drivers: new multiple audio / MIDI driver framework.
	*   Added ALSA PCM with mmap and capture support (stable).
	*   Added autodetection of ALSA PCM devices and PCM stream settings.
	*   Added ALSA Raw MIDI support (stable).
	*   Added Generic MIDI support (stable).
	*   Added OSS MIDI support (untested).

src/wave.h
/*****************************************************************************
 *
 * wave.h
 *
 * PHASEX:  [P]hase [H]armonic [A]dvanced [S]ynthesis [EX]periment
 *
 * Copyright (C) 1999-2015 Willaim Weston <william.h.weston@gmail.com>
 *
 * PHASEX is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * PHASEX is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with PHASEX.  If not, see <http://www.gnu.org/licenses/>.
 *
 *****************************************************************************/
#ifndef _PHASEX_WAVE_H_
#define _PHASEX_WAVE_H_

#include "phasex.h"


/* hardcoded waveform numbers */
#define WAVE_SINE           0
#define WAVE_TRIANGLE       1
#define WAVE_SAW            2
#define WAVE_REVSAW         3
#define WAVE_SQUARE         4
#define WAVE_STAIR          5
#define WAVE_SAW_S          6
#define WAVE_REVSAW_S       7
#define WAVE_SQUARE_S       8
#define WAVE_STAIR_S        9
#define WAVE_JUNO_OSC       10
#define WAVE_JUNO_SAW       11
#define WAVE_JUNO_SQUARE    12
#define WAVE_JUNO_POLY      13
#define WAVE_ANALOG_SINE_1  14
#define WAVE_ANALOG_SINE_2  15
#define WAVE_ANALOG_SQUARE  16
#define WAVE_VOX_1          17
#define WAVE_VOX_2          18
#define WAVE_POLY_SINE      19
#define WAVE_POLY_SAW       20
#define WAVE_POLY_REVSAW    21
#define WAVE_POLY_SQUARE_1  22
#define WAVE_POLY_SQUARE_2  23
#define WAVE_POLY_1         24
#define WAVE_POLY_2         25
#define WAVE_POLY_3         26
#define WAVE_POLY_4         27
#define WAVE_NULL           28
#define WAVE_IDENTITY       29


/* lookup macro for exponential frequency scaling */
#define halfsteps_to_freq_mult(val)                                     \
	(freq_shift_table[(int)(((val)*F_TUNING_RESOLUTION)+FREQ_SHIFT_ZERO_OFFSET)])

/* oscillator wave table */
#ifdef MALLOC_WAVE_TABLE

#define wave_get(wave_num, sample_num)                                  \
	(wave_table[((((unsigned int)(wave_num)) * (WAVEFORM_SIZE + 4)) + ((unsigned int)(sample_num)))])
#define wave_lookup(wave_num, sample_num)                               \
	(wave_table[((((unsigned int)(wave_num)) * (WAVEFORM_SIZE + 4)) + ((unsigned int)(sample_num)))])
#define wave_set(wave_num, sample_num, value)                           \
	wave_table[((((unsigned int)(wave_num)) * (WAVEFORM_SIZE + 4)) + ((unsigned int)(sample_num)))] = (sample_t)(value)
#define wave_inc(wave_num, sample_num, value)                           \
	wave_table[((((unsigned int)(wave_num)) * (WAVEFORM_SIZE + 4)) + ((unsigned int)(sample_num)))] += (sample_t)(value)

extern sample_t *wave_table;

#else

#define wave_get(wave_num, sample_num)                                  \
	(osc_table[(unsigned int)(wave_num)][(unsigned int)(sample_num)])
#define wave_lookup(wave_num, sample_num)                               \
	(osc_table[(unsigned int)(wave_num)][(unsigned int)(sample_num)])
#define wave_set(wave_num, sample_num, value)                           \
	osc_table[(unsigned int)(wave_num)][(unsigned int)(sample_num)] = (sample_t)(value)
#define wave_inc(wave_num, sample_num, value)                           \
	osc_table[wave_num][sample_num] += (sample_t)(value)

extern sample_t osc_table[NUM_WAVEFORMS][WAVEFORM_SIZE + 4];

#endif


/* lookup tables */
extern sample_t freq_shift_table[FREQ_SHIFT_TABLE_SIZE];
extern sample_t freq_table[128][648];
extern sample_t keyfollow_table[128][128];
extern sample_t mix_table[128];
extern sample_t pan_table[128];
extern sample_t gain_table[128];
extern sample_t velocity_gain_table[128][128];
extern sample_t env_curve[ENV_CURVE_SIZE + 1];
extern int      env_interval_dur[6][128];
extern int      env_table[128];

/* global base tuning frequency */
extern double   a4freq;


void build_freq_table(void);
void build_freq_shift_table(void);
void build_waveform_tables(void);
void build_env_tables(void);
void build_mix_table(void);
void build_pan_table(void);
void build_gain_table(void);
void build_velocity_gain_table(void);
void build_keyfollow_table(void);

#ifdef ENABLE_SAMPLE_LOADING
int load_waveform_sample(int wavenum, int num_cycles, double octaves, sample_t scale);
#endif /* ENABLE_SAMPLE_LOADING */

#ifdef NEED_GENERIC_HERMITE
sample_t hermite(sample_t *buf, unsigned int max, sample_t sample_index);
#endif
sample_t chorus_hermite(sample_t *buf, sample_t sample_index);
sample_t osc_table_hermite(int wave_num, sample_t sample_index);
sample_t osc_table_linear(int wave_num, sample_t sample_index);


/* these are the functions for building the initial waveforms */
/* these should _only_ be used in the startup code */
sample_t func_square(unsigned int sample_num);
sample_t func_saw(unsigned int sample_num);
sample_t func_revsaw(unsigned int sample_num);
sample_t func_saw_s(unsigned int sample_num);
sample_t func_revsaw_s(unsigned int sample_num);
sample_t func_stair(unsigned int sample_num);
sample_t func_triangle(unsigned int sample_num);


#endif /* _PHASEX_WAVE_H_ */
